src/signature.cpp

```diff
@@ -0,0 +1,209 @@
+                                          _getMDPtrFromDigestType(ctx.getMaskingFunction()));
+        }
+
+        // This determines the buffer length
+        _EVP_PKEY_sign(keyCtx.get(),
+                       nullptr,
+                       &siglen,
+                       reinterpret_cast<const unsigned char *>(&messageDigest[0]),
+                       messageDigest.size());
+
+        _CRYPTO_malloc_init();
+        sig.reset(static_cast<unsigned char *>(_OPENSSL_malloc(siglen)));
+        _EVP_PKEY_sign(keyCtx.get(),
+                       sig.get(),
+                       &siglen,
+                       reinterpret_cast<const unsigned char *>(&messageDigest[0]),
+                       messageDigest.size());
+    }
+    catch (const OpenSSLException &e) {
+        throw MoCOCrWException(e.what());
+    }
+
+    return std::vector<uint8_t>(sig.get(), sig.get() + siglen);
+}
+
+std::vector<uint8_t> SignatureUtils::create(const AsymmetricPrivateKey &privateKey,
+                                            const std::string message)
+{
+    Context<AsymmetricPrivateKey> ctx(privateKey);
+
+    return create(ctx, message);
+}
+
+std::vector<uint8_t> SignatureUtils::create(const AsymmetricPrivateKey &privateKey,
+                                            std::vector<uint8_t> messageDigest)
+{
+    Context<AsymmetricPrivateKey> ctx(privateKey);
+
+    return create(ctx, messageDigest);
+}
+
+void SignatureUtils::verify(Context<AsymmetricPublicKey> &ctx,
+                            const std::vector<uint8_t> signature,
+                            const std::string message)
+{
+    std::vector<uint8_t> messageDigest;
+
+    try {
+        messageDigest = digestMessage(message, ctx.getHashingFunction());
+    }
+    catch (const OpenSSLException &e) {
+        throw MoCOCrWException(e.what());
+    }
+    catch (const MoCOCrWException &e) {
+        throw e;
+    }
+
+    try {
+        verify(ctx, signature, messageDigest);
+    }
+    catch (const OpenSSLException &e) {
+        throw MoCOCrWException(e.what());
+    }
+}
+
+void SignatureUtils::verify(Context<AsymmetricPublicKey> &ctx,
+                            const std::vector<uint8_t> signature,
+                            const std::vector<uint8_t> messageDigest)
+{
+    const auto keyCtx = _EVP_PKEY_CTX_new(ctx.getKey().internal());
+
+    try {
+        _EVP_PKEY_verify_init(keyCtx.get());
+        _EVP_PKEY_CTX_set_rsa_padding(keyCtx.get(), static_cast<int>(ctx.getPadding()));
+        _EVP_PKEY_CTX_set_signature_md(keyCtx.get(),
+                                       _getMDPtrFromDigestType(ctx.getHashingFunction()));
+
+        if (RSAPaddingMode::PSS == ctx.getPadding()) {
+            _EVP_PKEY_CTX_set_rsa_pss_saltlen(keyCtx.get(), ctx.getSaltLength());
+            _EVP_PKEY_CTX_set_rsa_mgf1_md(keyCtx.get(),
+                                          _getMDPtrFromDigestType(ctx.getMaskingFunction()));
+        }
+
```

Comment:
I wonder if this piece of code can be refactored into something like EVP_PKEY_CTX* createSigningContext() and reused in create() and verify().